libs/client/src/streaming.ts

```diff
@@ -0,0 +1,251 @@
+import { createParser } from 'eventsource-parser';
+import { getTemporaryAuthToken } from './auth';
+import { buildUrl } from './function';
+import { ApiError, defaultResponseHandler } from './response';
+import { storageImpl } from './storage';
+
+/**
+ * The stream API options. It requires the API input and also
+ * offers configuration options.
+ */
+type StreamOptions<Input> = {
+  /**
+   * The API input payload.
+   */
+  input: Input;
+
+  /**
+   * The maximum time interval in milliseconds between stream chunks. Defaults to 15s.
+   */
+  timeout?: number;
+
+  /**
+   * Whether it should auto-upload File-like types to fal's storage
+   * or not.
+   */
+  autoUpload?: boolean;
+};
+
+const EVENT_STREAM_TIMEOUT = 15 * 1000;
+
+type FalStreamEventType = 'message' | 'error' | 'done';
+
+type EventHandler = (event: any) => void;
+
+/**
+ * The class representing a streaming response. With t
+ */
+class FalStream<Input, Output> {
+  // properties
+  url: string;
+  options: StreamOptions<Input>;
+
+  // support for event listeners
+  private listeners: Map<FalStreamEventType, EventHandler[]> = new Map();
+  private buffer: Output[] = [];
+
+  // local state
+  private currentData: Output | undefined = undefined;
+  private lastEventTimestamp = 0;
+  private streamClosed = false;
+  private donePromise: Promise<Output>;
+
+  constructor(url: string, options: StreamOptions<Input>) {
+    this.url = url;
+    this.options = options;
+    this.donePromise = new Promise<Output>((resolve, reject) => {
+      if (this.streamClosed) {
+        reject(
+          new ApiError({
+            message: 'Streaming connection is already closed.',
+            status: 400,
+            body: undefined,
+          })
+        );
+      }
+      this.on('done', (data) => {
+        this.streamClosed = true;
+        resolve(data);
+      });
+      this.on('error', (error) => {
+        this.streamClosed = true;
+        reject(error);
+      });
```

Comment:
`done` and `error` are our conventions, right? Should we have special instructions on how to construct a streaming end point?

Reply:
they are, but also internal to our client impl. Out `Stream` type has events, the names I used were the same the Node.js read stream uses. The endpoint itself doesn't emit events, it responds with a streaming with each message 

As for instructions, our endpoint exposes standard HTTP streaming, so as long we made that clear, people should be able to just call it. I intend to upgrade our API pages to show streaming snippets